src/Core/BackendAggregator.vala

```diff
@@ -87,27 +87,41 @@ public class AppCenterCore.BackendAggregator : Backend, Object {
     }
 
     public Gee.Collection<Package> get_applications_for_category (AppStream.Category category) {
-        var apps = new Gee.TreeSet<Package> ((a, b) => {
-            return a.normalized_component_id.collate (b.normalized_component_id);
-        });
-
+        var apps = new Gee.HashMap<string, Package> ();
         foreach (var backend in backends) {
-            apps.add_all (backend.get_applications_for_category (category));
+            var results = backend.get_applications_for_category (category);
+
+            foreach (var result in results) {
+                if (apps.has_key (result.normalized_component_id)) {
+                    if (result.origin_score > apps[result.normalized_component_id].origin_score) {
+                        apps[result.normalized_component_id] = result;
+                    }
+                } else {
+                    apps[result.normalized_component_id] = result;
+                }
+            }
         }
 
-        return apps;
+        return apps.values;
     }
 
     public Gee.Collection<Package> search_applications (string query, AppStream.Category? category) {
-        var apps = new Gee.TreeSet<Package> ((a, b) => {
-            return a.normalized_component_id.collate (b.normalized_component_id);
-        });
-
+        var apps = new Gee.HashMap<string, Package> ();
         foreach (var backend in backends) {
-            apps.add_all (backend.search_applications (query, category));
+            var results = backend.search_applications (query, category);
+
+            foreach (var result in results) {
```

Comment:
It might be worth adding a:
```vala
var result_component_id = result.normalized_component_id;
```
And reusing it when necessary :slightly_smiling_face: